README
NFCTD - Netfilter Conntrack net-snmp statistics
===============================================

0. About
--------
This tools collects statistics about changes in netfilter-conntrack
and makes it accessible via snmp-agent.

nfctd lets you organize your statistics using bpf-filter - see
config/nfctd.conf.dist

1. Quick start
--------------
$> git clone <URL> nfctd
$> cd nfctd
$> autoreconf -i
$> ./configure --prefix=/opt/nfctd/ --sysconfdir=/opt/nfctd/etc
$> make && make install
$> systemctl daemon-reload
$> systemctl enable nfctd.service
$> systemctl start nfctd.service

2. Intergration with collectd
-----------------------------
Include collectd-nfctd.types.db in your collectd.conf:

    TypesDB     "/etc/nfctd/collectd-nfctd.types.db"

2.1 Query snmp
--------------

Example of collectd.conf snmp plugin section:

<Plugin snmp>
   <Data "nfctd">
       Type "nfctd"
       Table true
       Instance "RRZN-MIB::groupId"
       Values "RRZN-MIB::ipv4IcmpAssured" "RRZN-MIB::ipv4IcmpCount" "RRZN-MIB::ipv4IcmpHalfAssured" "RRZN-MIB::ipv4IcmpUnreplied" \
              "RRZN-MIB::ipv4TcpAssured" "RRZN-MIB::ipv4TcpCount" "RRZN-MIB::ipv4TcpHalfAssured" "RRZN-MIB::ipv4TcpStateClose" \
              "RRZN-MIB::ipv4TcpStateCloseWait" "RRZN-MIB::ipv4TcpStateEstablished" "RRZN-MIB::ipv4TcpStateFinWait" \
              "RRZN-MIB::ipv4TcpStateLastAck" "RRZN-MIB::ipv4TcpStateSynRecv" "RRZN-MIB::ipv4TcpStateSynSent" \
              "RRZN-MIB::ipv4TcpStateSynSentAgain" "RRZN-MIB::ipv4TcpStateTimeWait" "RRZN-MIB::ipv4TcpUnreplied" \
              "RRZN-MIB::ipv4UdpAssured" "RRZN-MIB::ipv4UdpCount" "RRZN-MIB::ipv4UdpHalfAssured" "RRZN-MIB::ipv4UdpUnreplied" \
              "RRZN-MIB::ipv6IcmpAssured" "RRZN-MIB::ipv6IcmpCount" "RRZN-MIB::ipv6IcmpHalfAssured" "RRZN-MIB::ipv6IcmpUnreplied" \
              "RRZN-MIB::ipv6TcpAssured" "RRZN-MIB::ipv6TcpCount" "RRZN-MIB::ipv6TcpHalfAssured" "RRZN-MIB::ipv6TcpStateClose" \
              "RRZN-MIB::ipv6TcpStateCloseWait" "RRZN-MIB::ipv6TcpStateEstablished" "RRZN-MIB::ipv6TcpStateFinWait" \
              "RRZN-MIB::ipv6TcpStateLastAck" "RRZN-MIB::ipv6TcpStateSynRecv" "RRZN-MIB::ipv6TcpStateSynSent" \
              "RRZN-MIB::ipv6TcpStateSynSentAgain" "RRZN-MIB::ipv6TcpStateTimeWait" "RRZN-MIB::ipv6TcpUnreplied" \
              "RRZN-MIB::ipv6UdpAssured" "RRZN-MIB::ipv6UdpCount" "RRZN-MIB::ipv6UdpHalfAssured" "RRZN-MIB::ipv6UdpUnreplied"
   </Data>

   <Host "localhost">
       Address "127.0.0.1"
       Version 1
       Community "public"
       Collect "nfctd"
       Interval 300
   </Host>
</Plugin>


debian/changelog
nfctd (0.1.0-1) unstable; urgency=low

  * Initial release

 -- Markus Kötter <koetter@luis.uni-hannover.de>  Wed, 02 Dec 2015 10:51:11 +0100


include/nfctd-snmp.h
/**
 * nfctd - netlinkfilter-conntrack net-snmp statictics
 *
 * Copyright (C) 2015 Rafal Lesniak & Markus Koetter
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef NFCTD_H
#define NFCTD_H

/*
 * function declarations
 */
void            init_nfctd(void);
void            initialize_table_nfctTable(void);
struct nfctTable_entry *nfctTable_get_by_groupId(u_long);
struct nfctTable_entry *nfctTable_createEntry(u_long);

Netsnmp_Node_Handler nfctTable_handler;
Netsnmp_First_Data_Point nfctTable_get_first_data_point;
Netsnmp_Next_Data_Point nfctTable_get_next_data_point;

/*
 * column number definitions for table nfctTable
 */
#define COLUMN_GROUPID		1
#define COLUMN_IPV4TCPCOUNT		2
#define COLUMN_IPV4TCPASSURED		3
#define COLUMN_IPV4TCPHALFASSURED		4
#define COLUMN_IPV4TCPUNREPLIED		5
#define COLUMN_IPV4TCPSTATESYNSENT		6
#define COLUMN_IPV4TCPSTATESYNRECV		7
#define COLUMN_IPV4TCPSTATEESTABLISHED		8
#define COLUMN_IPV4TCPSTATEFINWAIT		9
#define COLUMN_IPV4TCPSTATECLOSEWAIT		10
#define COLUMN_IPV4TCPSTATELASTACK		11
#define COLUMN_IPV4TCPSTATETIMEWAIT		12
#define COLUMN_IPV4TCPSTATECLOSE		13
#define COLUMN_IPV4TCPSTATESYNSENTAGAIN		14
#define COLUMN_IPV4UDPCOUNT		15
#define COLUMN_IPV4UDPASSURED		16
#define COLUMN_IPV4UDPHALFASSURED		17
#define COLUMN_IPV4UDPUNREPLIED		18
#define COLUMN_IPV4ICMPCOUNT		19
#define COLUMN_IPV4ICMPASSURED		20
#define COLUMN_IPV4ICMPHALFASSURED		21
#define COLUMN_IPV4ICMPUNREPLIED		22
#define COLUMN_IPV6TCPCOUNT		23
#define COLUMN_IPV6TCPASSURED		24
#define COLUMN_IPV6TCPHALFASSURED		25
#define COLUMN_IPV6TCPUNREPLIED		26
#define COLUMN_IPV6TCPSTATESYNSENT		27
#define COLUMN_IPV6TCPSTATESYNRECV		28
#define COLUMN_IPV6TCPSTATEESTABLISHED		29
#define COLUMN_IPV6TCPSTATEFINWAIT		30
#define COLUMN_IPV6TCPSTATECLOSEWAIT		31
#define COLUMN_IPV6TCPSTATELASTACK		32
#define COLUMN_IPV6TCPSTATETIMEWAIT		33
#define COLUMN_IPV6TCPSTATECLOSE		34
#define COLUMN_IPV6TCPSTATESYNSENTAGAIN		35
#define COLUMN_IPV6UDPCOUNT		36
#define COLUMN_IPV6UDPASSURED		37
#define COLUMN_IPV6UDPHALFASSURED		38
#define COLUMN_IPV6UDPUNREPLIED		39
#define COLUMN_IPV6ICMPCOUNT		40
#define COLUMN_IPV6ICMPASSURED		41
#define COLUMN_IPV6ICMPHALFASSURED		42
#define COLUMN_IPV6ICMPUNREPLIED		43
#define COLUMN_BPFFILTER		44
#define COLUMN_LABEL		45


/*
 * Typical data structure for a row entry
 */
struct nfctTable_entry {
  /*
   * Index values
   */
  u_long          groupId;

  /*
   * Column values
   */
  u_long          ipv4TcpCount;
  u_long          ipv4TcpAssured;
  u_long          ipv4TcpHalfAssured;
  u_long          ipv4TcpUnreplied;
  u_long          ipv4TcpStateSynSent;
  u_long          ipv4TcpStateSynRecv;
  u_long          ipv4TcpStateEstablished;
  u_long          ipv4TcpStateFinWait;
  u_long          ipv4TcpStateCloseWait;
  u_long          ipv4TcpStateLastAck;
  u_long          ipv4TcpStateTimeWait;
  u_long          ipv4TcpStateClose;
  u_long          ipv4TcpStateSynSentAgain;
  u_long          ipv4UdpCount;
  u_long          ipv4UdpAssured;
  u_long          ipv4UdpHalfAssured;
  u_long          ipv4UdpUnreplied;
  u_long          ipv4IcmpCount;
  u_long          ipv4IcmpAssured;
  u_long          ipv4IcmpHalfAssured;
  u_long          ipv4IcmpUnreplied;
  u_long          ipv6TcpCount;
  u_long          ipv6TcpAssured;
  u_long          ipv6TcpHalfAssured;
  u_long          ipv6TcpUnreplied;
  u_long          ipv6TcpStateSynSent;
  u_long          ipv6TcpStateSynRecv;
  u_long          ipv6TcpStateEstablished;
  u_long          ipv6TcpStateFinWait;
  u_long          ipv6TcpStateCloseWait;
  u_long          ipv6TcpStateLastAck;
  u_long          ipv6TcpStateTimeWait;
  u_long          ipv6TcpStateClose;
  u_long          ipv6TcpStateSynSentAgain;
  u_long          ipv6UdpCount;
  u_long          ipv6UdpAssured;
  u_long          ipv6UdpHalfAssured;
  u_long          ipv6UdpUnreplied;
  u_long          ipv6IcmpCount;
  u_long          ipv6IcmpAssured;
  u_long          ipv6IcmpHalfAssured;
  u_long          ipv6IcmpUnreplied;
  char            bpfFilter[255];
  size_t          bpfFilter_len;
  char            label[255];
  size_t          label_len;

  /*
   * Illustrate using a simple linked list
   */
  int             valid;
  struct nfctTable_entry *next;
};


#endif                          /* NFCTD_H */


src/nfctd.c
/**
 * nfctd - netlinkfilter-conntrack net-snmp statictics
 *
 * Copyright (C) 2015 Rafal Lesniak & Markus Koetter
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include <fcntl.h>
#include <errno.h>
#include <time.h>

#include <execinfo.h>

#include <ev.h>

#include <libmnl/libmnl.h>
#include <linux/netfilter/nf_conntrack_tcp.h>
#include <libnetfilter_conntrack/libnetfilter_conntrack.h>

#include <pcap/pcap.h>
#include <pcap/bpf.h>

#include <net-snmp/net-snmp-config.h>
#include <net-snmp/net-snmp-includes.h>
#include <net-snmp/agent/net-snmp-agent-includes.h>

#include <netinet/ip6.h>
#include <netinet/ip.h>
#define __FAVOR_BSD
#include <netinet/tcp.h>
#include <netinet/udp.h>
#undef __FAVOR_BSD

#include "nfctd-snmp.h"
#include "lcfg_static.h"

#define IP6_HDRLEN 40  // IPv6 header length
#define TCP_HDRLEN 20  // TCP header length, excludes options data
#define UDP_HDRLEN 8
#define PKT_MAX_SIZE 256
#define ONE_MILLION 1000000

struct nfctd {
	struct lcfg *config;
	struct ev_loop *loop;
	struct nfct_handle *netlink_handle;
	struct nfctd_group *groups;
	ev_io nf_callback_watcher;
	ev_io *iow;
	ev_timer snmp_timer;
	ev_check snmp_check;
	ev_prepare snmp_prepare;
	int netlink_handle_fd;
	int snmp_numfds;
	int group_count;
	fd_set fdset;
};

struct nfctd_group {
	char *filter;
	char *label;
	int id;
	struct bpf_program *bpf;
	struct nfctd_group *next;
};

struct nfctd *g_nfctd;

volatile sig_atomic_t done = 0;

ev_signal signal_watcher;

#define MAX(a, b) \
	( {__typeof__( a )_a = ( a ); \
	   __typeof__( b )_b = ( b ); \
	   _a > _b ? _a : _b; } )

#define MIN(a, b)              \
	( {__typeof__( a )_a = ( a ); \
	   __typeof__( b )_b = ( b ); \
	   _a < _b ? _a : _b; } )

/**
 * TCP counting function
 */
void __handle_tcp_v4(struct nf_conntrack *ct, struct nfctTable_entry *entry)
{

	switch( nfct_get_attr_u8(ct, ATTR_TCP_STATE) )
	{

	case TCP_CONNTRACK_SYN_SENT:
		entry->ipv4TcpStateSynSent++;
		break;
	case TCP_CONNTRACK_SYN_RECV:
		entry->ipv4TcpStateSynRecv++;
		break;
	case TCP_CONNTRACK_ESTABLISHED:
		entry->ipv4TcpStateEstablished++;
		break;
	case TCP_CONNTRACK_FIN_WAIT:
		entry->ipv4TcpStateFinWait++;
		break;
	case TCP_CONNTRACK_LAST_ACK:
		entry->ipv4TcpStateLastAck++;
		break;
	case TCP_CONNTRACK_TIME_WAIT:
		entry->ipv4TcpStateTimeWait++;
		break;
	case TCP_CONNTRACK_CLOSE:
		entry->ipv4TcpStateClose++;
		break;
	case TCP_CONNTRACK_CLOSE_WAIT:
		entry->ipv4TcpStateCloseWait++;
		break;
	case TCP_CONNTRACK_SYN_SENT2:
		entry->ipv4TcpStateSynSentAgain++;
		break;
	case TCP_CONNTRACK_NONE:
		break;
	default:
		fprintf(stderr, "%s:%d: tcp default case (%d)\n", __FUNCTION__, __LINE__,
		        nfct_get_attr_u8(ct, ATTR_TCP_STATE) );
	}
}

/**
 * TCPv6 counting function
 */
void __handle_tcp_v6(struct nf_conntrack *ct, struct nfctTable_entry *entry)
{

	switch( nfct_get_attr_u8(ct, ATTR_TCP_STATE) )
	{

	case TCP_CONNTRACK_SYN_SENT:
		entry->ipv6TcpStateSynSent++;
		break;
	case TCP_CONNTRACK_SYN_RECV:
		entry->ipv6TcpStateSynRecv++;
		break;
	case TCP_CONNTRACK_ESTABLISHED:
		entry->ipv6TcpStateEstablished++;
		break;
	case TCP_CONNTRACK_FIN_WAIT:
		entry->ipv6TcpStateFinWait++;
		break;
	case TCP_CONNTRACK_LAST_ACK:
		entry->ipv6TcpStateLastAck++;
		break;
	case TCP_CONNTRACK_TIME_WAIT:
		entry->ipv6TcpStateTimeWait++;
		break;
	case TCP_CONNTRACK_CLOSE:
		entry->ipv6TcpStateClose++;
		break;
	case TCP_CONNTRACK_CLOSE_WAIT:
		entry->ipv6TcpStateCloseWait++;
		break;
	case TCP_CONNTRACK_SYN_SENT2:
		entry->ipv6TcpStateSynSentAgain++;
		break;
	case TCP_CONNTRACK_NONE:
		break;
	default:
		fprintf(stderr, "%s:%d: tcp default case (%d)\n", __FUNCTION__, __LINE__,
		        nfct_get_attr_u8(ct, ATTR_TCP_STATE) );
	}
}

void __l3_inet6(struct nf_conntrack *ct, struct nfctTable_entry *entry)
{
	u_int8_t l4proto;
	u_int32_t status;

	l4proto = nfct_get_attr_u8(ct, ATTR_ORIG_L4PROTO);
	status = nfct_get_attr_u32(ct, ATTR_STATUS);

	switch( l4proto )
	{
	case IPPROTO_TCP:
		entry->ipv6TcpCount++;

		if( status & IPS_ASSURED )
		{
			entry->ipv6TcpAssured++;
		} else
		if( !( status & IPS_SEEN_REPLY ) )
		{
			entry->ipv6TcpUnreplied++;
		} else
		{
			entry->ipv6TcpHalfAssured++;
		}

		__handle_tcp_v6(ct, entry);
		break;

	case IPPROTO_UDP:
		entry->ipv6UdpCount++;

		if( status & IPS_ASSURED )
		{
			entry->ipv6UdpAssured++;
		} else
		if( !( status & IPS_SEEN_REPLY ) )
		{
			entry->ipv6UdpUnreplied++;
		} else
		{
			entry->ipv6UdpHalfAssured++;
		}

	case IPPROTO_ICMPV6:
		entry->ipv6IcmpCount++;

		if( status & IPS_ASSURED )
		{
			entry->ipv6IcmpAssured++;
		} else
		if( !( status & IPS_SEEN_REPLY ) )
		{
			entry->ipv6IcmpUnreplied++;
		} else
		{
			entry->ipv6IcmpHalfAssured++;
		}

		break;

	default:
		fprintf(stderr, "%s:%d: default %d", __FUNCTION__, __LINE__, l4proto);
	}
}

void __l3_inet(struct nf_conntrack *ct, struct nfctTable_entry *entry)
{
	u_int8_t l4proto;
	u_int32_t status;

	l4proto = nfct_get_attr_u8(ct, ATTR_ORIG_L4PROTO);
	status = nfct_get_attr_u32(ct, ATTR_STATUS);

	switch( l4proto )
	{

	case IPPROTO_TCP:
		entry->ipv4TcpCount++;

		if( status & IPS_ASSURED )
		{
			entry->ipv4TcpAssured++;
		} else
		if( !( status & IPS_SEEN_REPLY ) )
		{
			entry->ipv4TcpUnreplied++;
		} else
		{
			entry->ipv4TcpHalfAssured++;
		}

		__handle_tcp_v4(ct, entry);
		break;

	case IPPROTO_UDP:
		entry->ipv4UdpCount++;

		if( status & IPS_ASSURED )
		{
			entry->ipv4UdpAssured++;
		} else
		if( !( status & IPS_SEEN_REPLY ) )
		{
			entry->ipv4UdpUnreplied++;
		} else
		{
			entry->ipv4UdpHalfAssured++;
		}

		break;

	case IPPROTO_ICMP:
		entry->ipv4IcmpCount++;

		if( status & IPS_ASSURED )
		{
			entry->ipv4IcmpAssured++;
		} else
		if( !( status & IPS_SEEN_REPLY ) )
		{
			entry->ipv4IcmpUnreplied++;
		} else
		{
			entry->ipv4IcmpHalfAssured++;
		}

		break;

	default:
		fprintf(stderr, "%s:%d: default %d", __FUNCTION__, __LINE__, l4proto);
	}
}

void increment_count_by_group_id(struct nf_conntrack *ct, struct nfctTable_entry *entry)
{
	u_int8_t l3proto;

	l3proto = nfct_get_attr_u8(ct, ATTR_ORIG_L3PROTO);

	switch( l3proto )
	{
	case AF_INET:
		__l3_inet(ct, entry);
		break;
	case AF_INET6:
		__l3_inet6(ct, entry);
		break;
	default:
		fprintf(stderr, "%s:%d : l3proto default case: %d\n", __FUNCTION__, __LINE__, l3proto);
		break;
	}
}

int nfctsa_data_cb(const struct nlmsghdr *nlh, int nf_conntrack_msg_type,  struct nf_conntrack *ct, void *data)
{
	struct nfctd_group *g;

	u_int8_t l4proto, l3proto;

	struct nfctTable_entry *entry;

	struct tcphdr *tcph;
	struct udphdr *udph;
	int len = 0;

	//Datagram to represent the packet
	char datagram[PKT_MAX_SIZE];
	memset(datagram, 0, PKT_MAX_SIZE);

	l3proto = nfct_get_attr_u8(ct, ATTR_ORIG_L3PROTO);
	l4proto = nfct_get_attr_u8(ct, ATTR_ORIG_L4PROTO);

	if( l3proto == AF_INET )
	{
		struct iphdr *iph = (struct iphdr *) datagram;

		//Fill in the IP Header
		iph->ihl = 5;
		iph->version = 4;
		iph->tos = 0;

		iph->check = 0;
		iph->saddr = nfct_get_attr_u32(ct, ATTR_ORIG_IPV4_SRC);
		iph->daddr = nfct_get_attr_u32(ct, ATTR_ORIG_IPV4_DST);

		switch( l4proto )
		{
		case IPPROTO_TCP:
			tcph = (struct tcphdr *) ( datagram + sizeof( struct ip ) );
			tcph->th_sport = nfct_get_attr_u16(ct, ATTR_ORIG_PORT_SRC);
			tcph->th_dport = nfct_get_attr_u16(ct, ATTR_ORIG_PORT_DST);

			iph->protocol = IPPROTO_TCP;
			len = iph->tot_len = sizeof( struct iphdr ) + sizeof( struct tcphdr );
			break;

		case IPPROTO_UDP:
			udph = (struct udphdr *) ( datagram + sizeof( struct ip ) );
			udph->uh_sport = nfct_get_attr_u16(ct, ATTR_ORIG_PORT_SRC);
			udph->uh_dport = nfct_get_attr_u16(ct, ATTR_ORIG_PORT_DST);
			iph->protocol = IPPROTO_UDP;
			len = iph->tot_len = sizeof( struct iphdr ) + sizeof( struct udphdr );
			break;

		case IPPROTO_ICMP:
			iph->protocol = IPPROTO_ICMP;
			len = iph->tot_len = sizeof( struct iphdr );
			break;

		default:
			fprintf(stderr, "%s: default %d", __FUNCTION__, l4proto);
		}
	}else
	if( l3proto == AF_INET6 )
	{
		struct ip6_hdr *iphdr = (struct ip6_hdr *) datagram;

		iphdr->ip6_flow = htonl( ( 6 << 28 ) | ( 0 << 20 ) | 0);
		iphdr->ip6_hops = 255;

		memcpy(&iphdr->ip6_src, nfct_get_attr(ct, ATTR_ORIG_IPV6_SRC), sizeof( uint32_t )*4);
		memcpy(&iphdr->ip6_dst, nfct_get_attr(ct, ATTR_ORIG_IPV6_DST), sizeof( uint32_t )*4);

		len = sizeof( struct ip6_hdr );

		switch( l4proto )
		{
		case IPPROTO_TCP:
			tcph = (struct tcphdr *) ( datagram + IP6_HDRLEN );
			iphdr->ip6_plen = htons(TCP_HDRLEN);
			iphdr->ip6_nxt = IPPROTO_TCP;
			tcph->th_sport = ntohs(nfct_get_attr_u16(ct, ATTR_ORIG_PORT_SRC) );
			tcph->th_dport = ntohs(nfct_get_attr_u16(ct, ATTR_ORIG_PORT_DST) );
			len = IP6_HDRLEN + TCP_HDRLEN;
			break;

		case IPPROTO_UDP:
			udph = (struct udphdr *) ( datagram + IP6_HDRLEN );
			udph->uh_sport = ntohs(nfct_get_attr_u16(ct, ATTR_ORIG_PORT_SRC) );
			udph->uh_dport = ntohs(nfct_get_attr_u16(ct, ATTR_ORIG_PORT_DST) );
			iphdr->ip6_nxt = IPPROTO_UDP;
			len = IP6_HDRLEN + UDP_HDRLEN;
			break;

		case IPPROTO_ICMPV6:
			len = IP6_HDRLEN;
			break;

		default:
			fprintf(stderr, "%s:%d: default %d", __FUNCTION__, __LINE__, l4proto);
		}
	}

	g = g_nfctd->groups;

	for(; g != NULL; g = g->next )
	{
		if( bpf_filter(g->bpf->bf_insns,
		               (const u_char *) &datagram, len, len) > 0 )
		{
			entry = (struct nfctTable_entry *) nfctTable_get_by_groupId(g->id);
			increment_count_by_group_id(ct, entry);
		}
	}

	return MNL_CB_OK;
}

static void nfct_ev_netlink_cb(EV_P_ ev_io *w, int revents)
{
	nfct_catch(g_nfctd->netlink_handle);
}

static void timeout_cb_snmp(EV_P_ ev_timer *w, int revents)
{
}

static void io_snmp_cb(EV_P_ ev_io *w, int revents)
{
}

static void snmp_prepare_cb(struct ev_loop *loop, ev_prepare *w, int revents)
{
	struct timeval ts;
	struct timeval ts2 = {
		1, 0
	};
	int block = 1;
	int snmp_numfds = 0;
	ev_tstamp ala = 0.0;
	/*
	   All permutations has been tested, this is the only one which seems to work
	 */
	snmp_select_info(&snmp_numfds, &g_nfctd->fdset, &ts2, &block);

	if( block == 0 )
	{

		if( ev_is_active(&g_nfctd->snmp_timer) )
			ev_timer_stop(loop, &g_nfctd->snmp_timer);

		memcpy(&ts, &ts2, sizeof( struct timeval ) );

		ts.tv_usec %= ONE_MILLION;
		ala = ts.tv_sec + ts.tv_usec * 1e-6;
		ala = MAX(ala, 1e-6);
		ala = MIN(ala, 15.0);

		ev_timer_init(&g_nfctd->snmp_timer, timeout_cb_snmp, ala, 0.);
		ev_timer_start(loop, &g_nfctd->snmp_timer);

	}

	for( int i = 0; i < g_nfctd->snmp_numfds; i++ )
		ev_io_stop(loop, g_nfctd->iow + i);

	if( snmp_numfds != g_nfctd->snmp_numfds )
	{
		free(g_nfctd->iow);
		g_nfctd->snmp_numfds = snmp_numfds;
		g_nfctd->iow = malloc(g_nfctd->snmp_numfds * sizeof( ev_io ) );
	}

	for( int i = 0; i < g_nfctd->snmp_numfds; i++ )
	{
		ev_io_init(g_nfctd->iow + i, io_snmp_cb, i, EV_READ);

		if( !FD_ISSET(i, &g_nfctd->fdset) )
			continue;

		ev_io_start(loop, g_nfctd->iow + i);
	}
}

// stop all watchers after blocking
static void snmp_check_cb(struct ev_loop *loop, ev_check *w, int revents)
{
	bool readable = false;

	for( int i = 0; i < g_nfctd->snmp_numfds; i++ )
	{
		if( !ev_is_active(g_nfctd->iow + i) )
			continue;

		readable |= ev_clear_pending(loop, g_nfctd->iow + i) & EV_READ;
		ev_io_stop(loop, g_nfctd->iow + i);
	}

	if( readable || ev_is_pending(&g_nfctd->snmp_timer) )
	{
		agent_check_and_process(0);

		if( ev_is_pending(&g_nfctd->snmp_timer) )
			snmp_timeout();
	}

	ev_timer_stop(loop, &g_nfctd->snmp_timer);

}

static void sigint_cb(struct ev_loop *loop, ev_signal *w, int revents)
{
	ev_break(loop, EVBREAK_ALL);
}

struct nfctd_group *nfctd_group_new(void)
{
	struct nfctd_group *ng;
	ng = malloc(sizeof( struct nfctd_group ) );
	memset(ng, 0x0, sizeof( struct nfctd_group ) );
	return ng;
}

void init_snmp_groups(void)
{
	struct nfctd_group *g;
	struct nfctTable_entry *entry;

	for( g = g_nfctd->groups; g; g = g->next )
	{
		entry = nfctTable_createEntry(g->id);
		memcpy(entry->bpfFilter, g->filter, strlen(g->filter) );
		entry->bpfFilter_len = strlen(g->filter);
		memcpy(entry->label, g->label, strlen(g->label) );
		entry->label_len = strlen(g->label);
	}
}

void init_groups(void)
{
	struct lcfgx_tree_node *groups;
	struct lcfgx_tree_node *root, *n;
	struct nfctd_group *grp;

	root = lcfgx_tree_new(g_nfctd->config);
	lcfgx_get_list(root, &groups, "groups");

	grp = nfctd_group_new();

	if( !g_nfctd->groups )
		g_nfctd->groups = grp;

	n = groups->value.elements;

	for(; n != NULL; n = n->next )
	{
		g_nfctd->group_count++;

		if( grp->label != NULL )
		{
			grp->next = nfctd_group_new();
			grp = grp->next;
		}

		struct pathhelper {
			const char *name;
			struct lcfgx_tree_node *node;
		} helper[3] =
		{{.name = "label"}, {.name = "filter"}, {.name = "id"}};

		for( int i=0; i < sizeof( helper )/sizeof( helper[0] ); i++ )
		{
			if( lcfgx_get_string(n, &helper[i].node, helper[i].name) != LCFGX_PATH_FOUND_TYPE_OK )
			{
				fprintf(stderr, "missing label \"%s\" \n", helper[i].name);
				exit(EXIT_FAILURE);
			}
		}

		grp->label = strndup(helper[0].node->value.string.data, helper[0].node->value.string.len);
		grp->filter = strndup(helper[1].node->value.string.data, helper[1].node->value.string.len);
		grp->id = atoi(helper[2].node->value.string.data);

		grp->bpf = malloc(sizeof( struct bpf_program ) );

		if( pcap_compile_nopcap(68, DLT_RAW, grp->bpf, grp->filter, 1, 0xffffffff) != 0 )
		{
			fprintf(stderr, "could not compile pcap filter \"%s\" for id %i \n", grp->filter, grp->id);
			exit(-1);
		}

		//fprintf(stderr, "group: id: %d label: '%s' filter: '%s'\n", grp->id, grp->label, grp->filter);
	}

	fprintf(stderr, "Loaded %d groups\n", g_nfctd->group_count);
}

void nfctd_prepare(void)
{
	// Get file descriptors
	g_nfctd->netlink_handle = nfct_open(CONNTRACK, NFCT_ALL_CT_GROUPS);

	if( g_nfctd->netlink_handle == NULL )
	{
		perror("nfct_open g_nfctd->netlink_handle");
		exit(EXIT_FAILURE);
	}

	g_nfctd->netlink_handle_fd = nfct_fd(g_nfctd->netlink_handle);
	fcntl(g_nfctd->netlink_handle_fd, F_SETFL, fcntl(g_nfctd->netlink_handle_fd, F_GETFL, 0) | O_NONBLOCK);

	if( nfct_callback_register2(g_nfctd->netlink_handle, NFCT_T_ALL, nfctsa_data_cb, NULL) < 0 )
	{
		perror("nfct_callback_register2 g_nfctd->netlink_handle");
		exit(EXIT_FAILURE);
	}
}

void sigsegv_backtrace_cb(int sig)
{
#define BACKTRACE_SIZE 256
	void *back[BACKTRACE_SIZE];
	size_t size;

	size = backtrace(back, BACKTRACE_SIZE);
	backtrace_symbols_fd(back, size, STDERR_FILENO);
	exit(-1);
}

void nfctd_new(const char *config_path)
{
	g_nfctd = malloc(sizeof( struct nfctd ) );
	memset(g_nfctd, 0x0, sizeof( struct nfctd ) );

	g_nfctd->loop = EV_DEFAULT;
	g_nfctd->config = lcfg_new(config_path);

	if( lcfg_parse(g_nfctd->config) != lcfg_status_ok )
	{
		fprintf(stderr, "lcfg error: %s\n", lcfg_error_get(g_nfctd->config) );
		exit(EXIT_FAILURE);
	}

	nfctd_prepare();
	init_groups();
	init_snmp_groups();

	ev_signal_init(&signal_watcher, sigint_cb, SIGINT | SIGTERM);
	ev_signal_start(g_nfctd->loop, &signal_watcher);

	ev_io_init(&g_nfctd->nf_callback_watcher, nfct_ev_netlink_cb,
	           g_nfctd->netlink_handle_fd, EV_READ);

	ev_timer_init(&g_nfctd->snmp_timer, 0, 1., 0.);
	ev_prepare_init(&g_nfctd->snmp_prepare, snmp_prepare_cb);
	ev_prepare_start(g_nfctd->loop, &g_nfctd->snmp_prepare);

	ev_check_init(&g_nfctd->snmp_check, snmp_check_cb);
	ev_check_start(g_nfctd->loop, &g_nfctd->snmp_check);

	ev_io_start(g_nfctd->loop, &g_nfctd->nf_callback_watcher);
	signal(SIGSEGV, sigsegv_backtrace_cb);
}

/**
 * Main loop
 */
int main(int argc, char *argv[])
{

	if( argc != 2 )
	{
		fprintf(stderr, "usage: %s FILE\n", argv[0]);
		return -1;
	}

	nfctd_new(argv[1]);

	snmp_enable_stderrlog();
	snmp_set_do_debugging(0);

	netsnmp_ds_set_boolean(NETSNMP_DS_APPLICATION_ID, NETSNMP_DS_AGENT_ROLE, 1);

	/* initialize the agent library */
	init_agent("nfctd-agent");
	init_nfctd();

	init_snmp("nfctd");
	snmp_log(LOG_INFO, "nfctd is up and running PID=(%d).\n", getpid() );

	agent_check_and_process(0); // 0 non-blocking

	ev_run(g_nfctd->loop, 0);

	snmp_log(LOG_INFO, "exiting...\n");

	snmp_shutdown("nfctd");
	nfct_close(g_nfctd->netlink_handle);

	return 0;
}
